linter: react/exhaustive-deps false negative

### What version of Oxlint are you using?

0.15.15

### What command did you run?

`oxlint ./src`

### What does your `.oxlintrc.json` config file look like?

```jsonc
{
    "$schema": "./node_modules/oxlint/configuration_schema.json",
    "rules": {
        "react/exhaustive-deps": "warn",
        "no-unused-vars": "off"
    },
    "env": {
        "builtin": true,
        "browser": true
    },
    "ignorePatterns": ["dist/**", "out/**", "node_modules/**"]
}
```


### What happened?

For this code:
```js
import { useCallback, useEffect } from "react";

function Component({ foo }) {
    const log = useCallback(() => {
        console.log(foo);
    }, [foo]);
    useEffect(() => {
        log();
    }, []);
}
```
no warning is emitted.

Both `react-hooks/exhaustive-deps` from `eslint-plugin-react-hooks` and `lint/correctness/useExhaustiveDependencies` from `biome` emit a warning:

`react-hooks/exhaustive-deps`:
```
/home/berni/Documents/src/test/linters/src/index.js
  9:8  warning  React Hook useEffect has a missing dependency: 'log'. Either include it or remove the dependency array  react-hooks/exhaustive-deps

✖ 1 problem (0 errors, 1 warning)
```

`lint/correctness/useExhaustiveDependencies`:
```
./src/index.js:7:5 lint/correctness/useExhaustiveDependencies ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ✖ This hook does not specify all of its dependencies: log
  
    5 │         console.log(foo);
    6 │     }, [foo]);
  > 7 │     useEffect(() => {
      │     ^^^^^^^^^
    8 │         log();
    9 │     }, []);
  
  ℹ This dependency is not specified in the hook dependency list.
  
     6 │     }, [foo]);
     7 │     useEffect(() => {
   > 8 │         log();
       │         ^^^
     9 │     }, []);
    10 │ }
  
  ℹ Either include it or remove the dependency array
  

Checked 1 file in 1396µs. No fixes applied.
Found 1 error.
lint ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ✖ Some errors were emitted while running checks.
```

--- a/crates/oxc_linter/src/rules/react/exhaustive_deps.rs
+++ b/crates/oxc_linter/src/rules/react/exhaustive_deps.rs
@@ -893,7 +893,7 @@ fn is_stable_value<'a, 'b>(
                 return false;
             };
 
-            if init_name == "useRef" || init_name == "useCallback" {
+            if init_name == "useRef" {
                 return true;
             }
 
@@ -2009,18 +2009,20 @@ fn test() {
             </>
           );
         }",
-        r"function Example() {
-          const foo = useCallback(() => {
-            foo();
-          }, []);
-        }",
-        r"function Example({ prop }) {
-          const foo = useCallback(() => {
-            if (prop) {
-              foo();
-            }
-          }, [prop]);
-        }",
+        // we don't support the following two cases as they would both cause an infinite loop at runtime
+        //  r"function Example() {
+        //    const foo = useCallback(() => {
+        //      foo();
+        //    }, []);
+        //  }",
+        //
+        //  r"function Example({ prop }) {
+        //    const foo = useCallback(() => {
+        //      if (prop) {
+        //        foo();
+        //      }
+        //    }, [prop]);
+        //  }",
         r"function Hello() {
           const [state, setState] = useState(0);
           useEffect(() => {
@@ -3324,21 +3326,20 @@ fn test() {
         r"function Thing() {
           useEffect(async () => {});
         }",
-        // TODO: not supported yet
+        // NOTE: intentionally not supported, as `foo` would be referenced before it's declaration
         // r"function Example() {
         //   const foo = useCallback(() => {
         //     foo();
         //     }, [foo]);
         //     }",
-        // TODO: not supported yet
-        // r"function Example({ prop }) {
-        //   const foo = useCallback(() => {
-        //     prop.hello(foo);
-        //   }, [foo]);
-        //   const bar = useCallback(() => {
-        //     foo();
-        //   }, [foo]);
-        // }",
+        r"function Example({ prop }) {
+          const foo = useCallback(() => {
+            prop.hello(foo);
+          }, [foo]);
+          const bar = useCallback(() => {
+            foo();
+          }, [foo]);
+        }",
         r"function MyComponent() {
           const local = {};
           function myEffect() {
@@ -3563,6 +3564,17 @@ fn test() {
             <></>
           )
         }",
+        // https://github.com/oxc-project/oxc/issues/9788
+        r#"import { useCallback, useEffect } from "react";
+
+        function Component({ foo }) {
+          const log = useCallback(() => {
+          console.log(foo);
+        }, [foo]);
+        useEffect(() => {
+          log();
+        }, []);
+        }"#,
     ];
 
     Tester::new(ExhaustiveDeps::NAME, ExhaustiveDeps::PLUGIN, pass, fail).test_and_snapshot();

